src/manifolds/StiefelSubmersionMetric.jl

```diff
@@ -0,0 +1,293 @@
+    Y,
+) where {n,k}
+    α = metric(M).α
+    T = typeof(one(Base.promote_eltypeof(p, X, Y, α)))
+    if n == k
+        return T(dot(X, Y)) / (2 * (α + 1))
+    elseif α == -1 // 2
+        return T(dot(X, Y))
+    else
+        return dot(X, Y) - (T(dot(p'X, p'Y)) * (2α + 1)) / (2 * (α + 1))
+    end
+end
+
+function inverse_retract!(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},<:StiefelSubmersionMetric},
+    X,
+    p,
+    q,
+    method::ShootingInverseRetraction{ExponentialRetraction},
+) where {n,k}
+    if k > div(n, 2)
+        shooting!(
+            M,
+            X,
+            p,
+            q,
+            method.retraction;
+            vector_transport=method.vector_transport,
+            initial_inverse_retraction=method.initial_inverse_retraction,
+            num_transport_points=method.num_transport_points,
+            tolerance=method.tolerance,
+            max_iterations=method.max_iterations,
+        )
+        return X
+    end
+    ts = range(0, 1; length=method.num_transport_points)
+    α = metric(M).α
+    M̂ = p'q
+    skewM̂ = project(SkewHermitianMatrices(k), M̂)
+    Q, N̂ = qr(q - p * M̂)
+    normN̂² = norm(N̂)^2
+    gap = sqrt(norm(M̂ - I)^2 + normN̂²) # γ
+    if iszero(gap)
+        fill!(X, false)
+        return X
+    end
+    c = gap / sqrt(norm(skewM̂)^2 + normN̂²)
+    A = rmul!(skewM̂, c)
+    R = c * N̂
+    S = allocate(A)
+    Aˢ = allocate(A)
+    Rˢ = allocate(R)
+    D = allocate(A)
+    C = allocate(A, Size(2k, 2k))
+    i = 1
+    while (gap > method.tolerance) && (i < method.max_iterations)
+        @views begin
+            C[1:k, 1:k] .= A ./ (α + 1)
+            C[1:k, (k + 1):(2k)] .= -R'
+            C[(k + 1):(2k), 1:k] .= R
+            fill!(C[(k + 1):(2k), (k + 1):(2k)], false)
+        end
+        D .= A .* (α / (α + 1))
+        @views begin
+            E = exp(C)[:, 1:k] * exp(D)
+            M = E[1:k, 1:k]
+            N = E[(k + 1):(2k), 1:k]
+        end
+        Aˢ .= M .- M̂
+        Rˢ .= N .- N̂
+        gap = sqrt(norm(Aˢ)^2 + norm(Rˢ)^2)
+        _vector_transport_factors!(Aˢ, Rˢ, S, M, N, gap, method.tolerance, Val(k))
+        for t in reverse(ts)[2:(end - 1)]
+            @views begin
+                E = exp(t * C)[:, 1:k] * exp(t * D)
+                M = E[1:k, 1:k]
+                N = E[(k + 1):(2k), 1:k]
+            end
+            _vector_transport_factors!(Aˢ, Rˢ, S, M, N, gap, method.tolerance, Val(k))
+        end
+        copyto!(M, I)
+        fill!(N, 0)
+        _vector_transport_factors!(Aˢ, Rˢ, S, M, N, gap, method.tolerance, Val(k))
+        A .-= Aˢ
+        R .-= Rˢ
+        i += 1
+    end
+    mul!(X, p, A)
+    mul!(X, Matrix(Q), R, true, true)
+    return X
+end
+
+function inverse_retract_project!(
+    M::MetricManifold{ℝ,<:Stiefel,<:StiefelSubmersionMetric},
+    X,
+    p,
+    q,
+)
+    X .= q .- p
+    project!(M, X, p, X)
+    return X
+end
+
+function _vector_transport_factors!(A, R, S, M, N, gap, ϵ, ::Val{k}) where {k}
```

Comment:
This isn't super human-readable unless we have the paper open. Should probably rewrite to explicitly show the points in terms of factors.

Reply:
I have opted to explicitly include the factorizations.